Dump operator names of a script module

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30467 Dump operator names of a script module**

Introduce function jit.export_opnames(module), which returns a list of all operator names used in the module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(torch.jit.export_opnames(m))

The outputs are in alphabetical order:
['aten::_convolution', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::append.Tensor', 'aten::cat', 'aten::dropout', 'aten::embedding', 'aten::matmul', 'aten::max.dim', 'aten::mul.Tensor', 'aten::permute', 'aten::relu', 'aten::t', 'aten::tanh', 'prim::ListConstruct', 'prim::TupleConstruct', 'prim::TupleUnpack']

Differential Revision: [D18801619](https://our.internmc.facebook.com/intern/diff/D18801619)

Commit:
Dump operator names of a script module

Commit:
Update on "Dump operator names of a script module"

ZolotukhinM commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -368,6 +369,37 @@ void Module::dump(
+std::vector<std::string> Module::opnames() const {
+  std::unordered_set<std::string> names;
+  export_opnames(*this, names);
+  return std::vector<std::string>(names.begin(), names.end());
```
Order of elements in the returned vector is non-deterministic since they come from `std::unordered_set`. That might result in hard to debug issues, so please make sure that the order is well-specified.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -222,6 +222,8 @@ struct TORCH_API Module : public Object {
       const std::string& filename,
       const ExtraFilesMap& extra_files = ExtraFilesMap()) const;
 
+  std::vector<std::string> opnames() const;
```
It doesn't have to be a method of `torch::jit::script::Module` class. I think it can and should live separately, and you actually already implemented it this way.

ZolotukhinM commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -368,6 +369,37 @@ void Module::dump(
+  auto methods = m.get_methods();
+  for (const auto& method : methods) {
+    const auto& func = method.function();
+    torch::jit::Code code(func.graph());
```
Why do we need to construct the `Code` object? Will traversing all nodes in the graph not be enough for some reason?

iseeyuan replied:
I think the code would contain the final ops for interpreter (when we bypass the optimizations in mobile). I'm not sure if emitting nodes would introduce more ops in future? Another reason is that I'm not sure how to know if a node is an operator node.

ZolotukhinM replied:
All exporters use nodes directly and hence this function should use that too. Also, `Code` doesn't do any optimizations, it just wraps some nodes with Instructions - it's an internal detail of how interpreter works and should not affect exporting.

ZolotukhinM commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -368,6 +369,37 @@ void Module::dump(
+namespace {
+void export_opnames(const script::Module& m, std::unordered_set<std::string>& opnames) {
+  auto methods = m.get_methods();
+  for (const auto& method : methods) {
```
Nit: `for (const auto& method : m.get_methods()) {`, no need in having `methods` variable.

ZolotukhinM reviewed (changes requested):
I think this doesn't have to be a method of Module, it can be a standalone pass. Please also find a couple of other comments inline.

Commit:
Update on "Dump operator names of a script module"


Introduce function script.module.opnames(), which returns a list of all operator names used in that module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(m.opnames())

Outputs:
['aten::append.Tensor', 'aten::addmm', 'aten::tanh', 'aten::add_.Tensor', 'aten::matmul', 'aten::relu', 'aten::cat', 'prim::TupleUnpack', 'prim::ListConstruct', 'aten::t', 'aten::_convolution', 'aten::mul.Tensor', 'aten::permute', 'aten::add.Tensor', 'aten::max.dim', 'aten::dropout', 'aten::embedding', 'prim::TupleConstruct']

Commit:
Update on "Dump operator names of a script module"


Introduce function jit.opnames(module), which returns a list of all operator names used in the module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(torch.jit.opnames(m))

The outputs are in alphabetical order:
['aten::_convolution', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::append.Tensor', 'aten::cat', 'aten::dropout', 'aten::embedding', 'aten::matmul', 'aten::max.dim', 'aten::mul.Tensor', 'aten::permute', 'aten::relu', 'aten::t', 'aten::tanh', 'prim::ListConstruct', 'prim::TupleConstruct', 'prim::TupleUnpack']

ZolotukhinM reviewed (commented):
Thanks, it looks better! I think we still should switch to use nodes directly (without going through `Code` - please see the comment inline) and also we need some tests for this.

Commit:
Update on "Dump operator names of a script module"


Introduce function jit.opnames(module), which returns a list of all operator names used in the module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(torch.jit.opnames(m))

The outputs are in alphabetical order:
['aten::_convolution', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::append.Tensor', 'aten::cat', 'aten::dropout', 'aten::embedding', 'aten::matmul', 'aten::max.dim', 'aten::mul.Tensor', 'aten::permute', 'aten::relu', 'aten::t', 'aten::tanh', 'prim::ListConstruct', 'prim::TupleConstruct', 'prim::TupleUnpack']

Commit:
Update on "Dump operator names of a script module"


Introduce function jit.opnames(module), which returns a list of all operator names used in the module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(torch.jit.opnames(m))

The outputs are in alphabetical order:
['aten::_convolution', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::append.Tensor', 'aten::cat', 'aten::dropout', 'aten::embedding', 'aten::matmul', 'aten::max.dim', 'aten::mul.Tensor', 'aten::permute', 'aten::relu', 'aten::t', 'aten::tanh', 'prim::ListConstruct', 'prim::TupleConstruct', 'prim::TupleUnpack']

iseeyuan commented:
Updated to use nodes in graph directly. @ZolotukhinM

Commit:
Update on "Dump operator names of a script module"


Introduce function jit.export_opnames(module), which returns a list of all operator names used in the module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(torch.jit.export_opnames(m))

The outputs are in alphabetical order:
['aten::_convolution', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::append.Tensor', 'aten::cat', 'aten::dropout', 'aten::embedding', 'aten::matmul', 'aten::max.dim', 'aten::mul.Tensor', 'aten::permute', 'aten::relu', 'aten::t', 'aten::tanh', 'prim::ListConstruct', 'prim::TupleConstruct', 'prim::TupleUnpack']

ZolotukhinM reviewed (approved):
This is good to go once you add tests, thanks!

Commit:
Update on "Dump operator names of a script module"


Introduce function jit.export_opnames(module), which returns a list of all operator names used in the module and its submodules. One usage is to have mobile custom build to link only operators in the returned list to save the mobile size. 

Example: 
import torch
m = torch.jit.load("example.pt")
print(torch.jit.export_opnames(m))

The outputs are in alphabetical order:
['aten::_convolution', 'aten::add.Tensor', 'aten::add_.Tensor', 'aten::addmm', 'aten::append.Tensor', 'aten::cat', 'aten::dropout', 'aten::embedding', 'aten::matmul', 'aten::max.dim', 'aten::mul.Tensor', 'aten::permute', 'aten::relu', 'aten::t', 'aten::tanh', 'prim::ListConstruct', 'prim::TupleConstruct', 'prim::TupleUnpack']
